src/main/java/com/mongodb/hibernate/jdbc/MongoResultSet.java

```diff
@@ -0,0 +1,282 @@
+/*
+ * Copyright 2025-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.hibernate.jdbc;
+
+/*
+ * Copyright 2025-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import static com.mongodb.hibernate.internal.MongoAssertions.assertNotNull;
+import static java.lang.String.format;
+
+import com.mongodb.client.MongoCursor;
+import com.mongodb.hibernate.internal.FeatureNotSupportedException;
+import java.math.BigDecimal;
+import java.sql.Array;
+import java.sql.Date;
+import java.sql.ResultSetMetaData;
+import java.sql.SQLException;
+import java.sql.Time;
+import java.sql.Timestamp;
+import java.util.Calendar;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Function;
+import org.bson.BsonDocument;
+import org.bson.BsonNull;
+import org.bson.BsonValue;
+import org.jspecify.annotations.Nullable;
+
+/**
+ * MongoDB Dialect's JDBC {@link java.sql.ResultSet} implementation class.
+ *
+ * <p>It only focuses on API methods Mongo Dialect will support. All the other methods are implemented by throwing
+ * exceptions in its parent {@link ResultSetAdapter adapter interface}.
+ */
+final class MongoResultSet implements ResultSetAdapter {
+
+    private class MongoResultSetMetadata implements ResultSetMetaDataAdapter {
+        @Override
+        public int getColumnCount() {
+            return fieldNames.size();
+        }
+
+        @Override
+        public String getColumnLabel(final int column) {
+            return fieldNames.get(column - 1);
+        }
+    }
+
+    private final MongoCursor<BsonDocument> mongoCursor;
+
+    private final List<String> fieldNames;
+
+    private @Nullable BsonDocument currentDocument;
+
+    private @Nullable BsonValue lastReadColumnValue;
```

Comment:
[optional]

We don't actually need to remember the last read `BsonValue`. All we need is to remember a `@Nullable Boolean`, telling us whether the last read column value was null. Maybe we should make that explicit and stop storing `BsonValue`?

Reply:
good point. changed.